Tests/NIOSOCKSTests/SOCKSServerHandshakeHandler+Tests.swift

```diff
@@ -0,0 +1,191 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2021 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+@testable import NIOSOCKS
+import XCTest
+
+class PromiseTestHandler: ChannelInboundHandler {
+    typealias InboundIn = ClientMessage
+    
+    let expectedGreeting: ClientGreeting
+    let greetingPromise: EventLoopPromise<Void>
+    let expectedRequest: SOCKSRequest
+    let requestPromise: EventLoopPromise<Void>
+    let expectedData: ByteBuffer
+    let dataPromise: EventLoopPromise<Void>
```

Comment:
Do these need to be promises, or can they be bools? If we're using embedded channel we presumably don't need the thread safety of promises, and it's much harder to ask the right questions if they're promises.

Reply:
Promises tends to be the NIO way, no? Bools seems nasty for this. If you're against promises then I'd rather use test expectations.

Reply:
Promises are the NIO way when concurrency is involved. Here this is EmbeddedChannel. It's either happened or it hasn't.